### DISCUSSION TITLE
Can we change joint configuration without reloading URDF?

### DISCUSSION THREAD

**Discussion by keiohta** [haosulab/SAPIEN#59] [Q&A] [answered]
Hi author, I would like to change a joint configuration without loading different URDFs. Can we do that?
For example, imagine we have a door object with a revolutionary joint on the right side (as the attached picture from [SAPIEN PartNet-Mobility Dataset browser](https://sapien.ucsd.edu/browse)), and we would like to change the joint to the left side of the door or change the joint to the prismatic joint without loading a different URDF file again.

I understand we can do that by changing the joint configuration (`origin`, `axis`, `type`, `limit`) described in URDF, but hopefully I would like to do it without reloading the file.

Thanks!

![image](https://user-images.githubusercontent.com/7234712/149616039-8ec9a487-ebd0-4289-b41b-bee8f6f473a3.png)

**Comment by fbxiang** [comment, marked as answer]
After an articulation is created, only the joint limit can be changed.
If you do not want to create multiple URDF files, an option is to recreate the URDF loader in python. SAPIEN allows building articulations from scratch, so you can always write your own URDF (or any other format) parser. (See https://sapien.ucsd.edu/docs/latest/tutorial/basic/create_articulations.html)
Also reloading files should not be a big problem since the slowest and memory intensive part (rendering resources) is cached and will only load once.

**Reply by keiohta** [reply]
Sure, I understand we can't change URDF configurations after loading them.

> Also reloading files should not be a big problem since the slowest and memory intensive part (rendering resources) is cached and will only load once.

Thanks for the tips. I actually measured the time as follows but haven't been able to see such performance gain by the cache. Here I attach the experimental code:
```python
import sapien.core as sapien

def main():
    from time import time_ns

    engine = sapien.Engine()
    scene = engine.create_scene()
    loader = scene.create_urdf_loader()

    urdfs = ["a.urdf",
             "b.urdf",
             "c.urdf",
             "d.urdf"]
    elapsed_times_ms = []
    for i in range(10):
        start_time_ns = time_ns()
        for urdf in urdfs:
            loader.load(urdf)
        elapsed_time_ms = (time_ns() - start_time_ns) / 10 ** 6
        elapsed_times_ms.append(f"itr={i} elapsed_time={elapsed_time_ms} ms")
    for str in elapsed_times_ms:
        print(str)


if __name__ == "__main__":
    main()
```

, and the result was

```
itr=0 elapsed_time=348.418888 ms
itr=1 elapsed_time=212.426163 ms
itr=2 elapsed_time=345.754236 ms
itr=3 elapsed_time=222.730418 ms
itr=4 elapsed_time=132.756032 ms
itr=5 elapsed_time=164.345239 ms
itr=6 elapsed_time=127.840826 ms
itr=7 elapsed_time=185.42149 ms
itr=8 elapsed_time=155.460915 ms
itr=9 elapsed_time=135.800838 ms
```

meaning that the performance gain hasn't been observed (if my understanding is correct, what you suggested was only `itr=0` takes long and `itr>0` should be shorter).
Could you take a look at the experimental code and let me know some problems if it has?

**Reply by fbxiang** [reply]
This looks right since you are not calling the "render" function, so no caching has happened in your code.
The loading in the renderer often take several seconds for detailed meshes so that is what we optimize.

**Reply by keiohta** [reply]
Oh, I see.
In the meantime, I found preparing multiple scenes can be one possible workaround. I'll ask again if it won't work and get back to your suggested method, and still have questions.
Thanks!!
